[generator] Fix bug 37175 and bug 44322

dalexsoto opened the pull request:
https://bugzilla.xamarin.com/show_bug.cgi?id=37175
https://bugzilla.xamarin.com/show_bug.cgi?id=44322

Sometimes iOS API is returning objects of the wrong type for example

- (NSURLSessionDownloadTask *)downloadTaskWithRequest:(NSURLRequest *)request

It clearly states that it will return an NSURLSessionDownloadTask
instance, but yet it returns a NSURLSessionTask and we throw an
InvalidCastException when creating the managed object. This is
fine in ObjC but for us in a type-safety context is not.

Introducing ForcedTypeAttribute when applying this attribute
we will use GetINativeObject instead of GetNSObject which
the former does not perfom a type check on the underlying object

API diff: https://gist.github.com/dalexsoto/0d2c7c551105e5bda803790fde62bb9d

monojenkins commented:
Build success

monojenkins commented:
Build success

monojenkins commented:
Build success

spouliot commented on docs/website/binding_types_reference_guide.md:
````diff
@@ -961,6 +961,41 @@ as your thread did not return control to the main loop. Uf your thread was some
+	[ForcedType]
+	NSUrlSessionDownloadTask CreateDownloadTask (NSUrlRequest request);
+}
+```
````
The `owns` parameter is not documented.

dalexsoto replied:
Fixed locally, good catch

dalexsoto commented:
I love the [return: ] idea, I already implemented this, also for trampolines on delegates, so I ended up with this restrictions

AttributeUsage (AttributeTargets.ReturnValue | AttributeTargets.Parameter | AttributeTargets.Property, AllowMultiple = false)]

Will add some generator tests and push again

dalexsoto commented:
@spouliot Implemented your requested changes, also updated the API diff in the description. `ForcedTypeAttribute` is only valid on `parameters`, `properties` and `return value`, added a generator test including properties, delegates, ref/out parameters and returns.

@rolfbjarne I would love your input too! This should cover [Bug 37175](https://bugzilla.xamarin.com/show_bug.cgi?id=37175) too.

monojenkins commented:
Build success

rolfbjarne commented on docs/website/binding_types_reference_guide.md:
```diff
@@ -961,6 +961,47 @@ as your thread did not return control to the main loop. Uf your thread was some
+definition.
+
+This is useful when the type described in a header does not match the returned type
+of the native method for example take the following Objective-C definition from `NSURLSession`:
```
missing comma (or period):

    of the native method, for example ...

dalexsoto replied:
Fixed! thanks!

rolfbjarne commented on docs/website/binding_types_reference_guide.md:
````diff
@@ -961,6 +961,47 @@ as your thread did not return control to the main loop. Uf your thread was some
+```
+
+The `ForcedTypeAttribute` also accepts a boolean value named `Owns` that is `false`
+by default `[ForcedType (owns: true)]`. The owns parameter could be used to follow
````
I think this is slightly more correct:

   The owns parameter _is_ used to ...

dalexsoto replied:
Fixed!

rolfbjarne commented on src/generator.cs:
```diff
@@ -4330,13 +4396,17 @@ void GenerateTypeLowering (MethodInfo mi, bool null_allowed_override, EnumMode e
+					by_ref_processing.AppendFormat("{0} = {0}Value != IntPtr.Zero ? Runtime.GetINativeObject<{1}> ({0}Value, {2}) : null;", pi.Name.GetSafeParamName (), RenderType (mai.Type.GetElementType ()), isForcedOwns);
 				} else {
-					by_ref_processing.AppendFormat("{0} = {0}Value != IntPtr.Zero ? Runtime.GetNSObject<{1}> ({0}Value) : null;", pi.Name.GetSafeParamName (), RenderType (mai.Type.GetElementType ()));
+					by_ref_processing.AppendFormat ("{0} = {0}Value != IntPtr.Zero ? Runtime.GetNSObject<{1}> ({0}Value) : null;", pi.Name.GetSafeParamName (), RenderType (mai.Type.GetElementType ()));
```
Only whitespace change?

dalexsoto replied:
heh that was my editor

dalexsoto replied:
Fixed!

monojenkins commented:
Build success

spouliot commented on src/generator.cs:
```diff
@@ -279,6 +279,47 @@ static void FillKeywordTable ()
+//
+// [1]: https://developer.apple.com/library/content/documentation/CoreFoundation/Conceptual/CFMemoryMgmt/Concepts/Ownership.html
+//
+
```
Let's not duplicate the documentation _anymore_, since it will be hard to ensure they stay consistent.
Just add a link to the documentation (so we know it's documented).
